staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go

```diff
@@ -203,45 +204,89 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
 			attemptsExpected: 0,
 			errExpected:      context.DeadlineExceeded,
 		},
+		{
+			name:      "context canceled before the second execution and immediate",
+			immediate: true,
+			context: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), time.Second)
+			},
+			callback: func(attempts int) (bool, error) {
+				return false, nil
+			},
+			attemptsExpected: 1,
+			errExpected:      context.DeadlineExceeded,
+			timer:            Backoff{Duration: 2 * time.Second}.Timer(),
+		},
+		{
+			name:      "immediate and long duration of condition and sliding false",
+			immediate: true,
+			sliding:   false,
+			context: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), time.Second)
+			},
+			callback: func(attempts int) (bool, error) {
+				if attempts >= 4 {
+					return true, nil
+				}
+				time.Sleep(time.Second / 5)
+				return false, nil
+			},
+			attemptsExpected: 4,
+			timer:            Backoff{Duration: time.Second / 5, Jitter: 0.001}.Timer(),
+		},
+		{
+			name:      "immediate and long duration of condition and sliding true",
+			immediate: true,
+			sliding:   true,
+			context: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), time.Second)
+			},
+			callback: func(attempts int) (bool, error) {
+				if attempts >= 4 {
+					return true, nil
+				}
+				time.Sleep(time.Second / 5)
```

Comment:
Are you trying to cause it to race?

Reply:
> Are you trying to cause it to race?

I explained my thoughts in [comment](https://github.com/kubernetes/kubernetes/pull/119762#issuecomment-1671778866). 

Basically, I created a callback function which requires a long duration to execute. So in a same duration, poll with sliding true/false will execute this function different times.  

If sliding is false, we'll keep execute this function and never waits, because timech is drained during the sleep in condition function. Otherwise, we'll execute less times because we'll have to wait a longer time before next execution.

Reply:
> How did you choose this amount to sleep?

Well, at first it was a big amount, and aojea suggests to reduce the time cost. So I tried to reduce it, but if this amount is too small, this test will fail in the CI(however it passed on my mac). I doubt it should be some accuracy problem. So I tried a few amounts, and this amount is ok on both sides(CI and my mac), and 1 second seems not too expensive.

Reply:
it took me a while to understand it too, I had to draw it actually, it plays with the periods and the attempts to make it predictable